dlib/cuda/cuda_dlib.cu

```diff
@@ -1720,10 +1762,34 @@ namespace dlib
             warp_reduce_atomic_add(*loss_out, loss);
         }
 
+    // ----------------------------------------------------------------------------------------
+
+        void compute_loss_binary_log_per_pixel::
+        do_work(
+            cuda_data_ptr<float> loss_cuda_work_buffer,
+            const float* truth_buffer,
+            const tensor& subnetwork_output,
+            tensor& gradient,
+            double& loss
+        )
+        {
+            CHECK_CUDA(cudaMemset(loss_cuda_work_buffer, 0, sizeof(float)));
+            sigmoid(gradient, subnetwork_output);
+
+            // The loss we output is the average loss over the mini-batch, and also over each element of the matrix output.
+            const double scale = 1.0 / (subnetwork_output.num_samples() * subnetwork_output.nr() * subnetwork_output.nc());
+
+            launch_kernel(_cuda_compute_loss_binary_log_per_pixel, max_jobs(gradient.size()),
+                loss_cuda_work_buffer.data(), gradient.device(), truth_buffer, subnetwork_output.device(), gradient.size(), scale);
+
+            float floss;
+            dlib::cuda::memcpy(&floss, loss_cuda_work_buffer);
+            loss = scale*floss;
+        }
 
         void compute_loss_multiclass_log_per_pixel::
         do_work(
-            float* loss_cuda_work_buffer,
+            cuda_data_ptr<float> loss_cuda_work_buffer,
             const uint16_t* truth_buffer,
```

Comment:
Can `truth_buffer` be a cuda_data_ptr too?  The way the code appears suggests `truth_buffer` is a host pointer.